tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -14161,6 +14161,97 @@ static bool IsUsageAttributeCompatible(AttributeList::Kind kind, bool &usageIn,
   return true;
 }
 
+bool DiagnoseRayQueryObject(QualType qt, Declarator &D, Sema &S) {
+  // now we know that qt is a TemplateSpecializationType
+  // handle diagnostics that relate to enumtypedecl vs literal first
+  const TemplateSpecializationType *TST =
+      dyn_cast<TemplateSpecializationType>(qt);
+  if (!TST) {
+    return false;
+  }
+
+  auto typeRecordDecl = qt->getAsCXXRecordDecl();
+  ClassTemplateSpecializationDecl *SpecDecl =
+      llvm::dyn_cast<ClassTemplateSpecializationDecl>(typeRecordDecl);
```

Comment:
It's not obvious to me what the type of `typeRecordDecl` is, and so I can't tell by reading just this code whether or not it is making an unnecessary copy or not.  An explicit `auto *` would help make that clearer, but I think really this could just be done inline (as I see it done in other places):

```suggestion
  ClassTemplateSpecializationDecl *SpecDecl =
      llvm::dyn_cast<ClassTemplateSpecializationDecl>(qt->getAsCXXRecordDecl());
```

Although in this case I think using auto would be ok because of the dyn_cast:


```suggestion
  auto *SpecDecl =
      llvm::dyn_cast<ClassTemplateSpecializationDecl>(qt->getAsCXXRecordDecl());
```